### PR TITLE
Add utility to reshard TFDS dataset for large scale training

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Add a utility to reshard TFDS dataset for large scale training, b/508024540.

 * Eliminates overhead of dynamic re-sharding during training by pre-sharding data to match the host count.
 * Tests show a reduction in training loss (from 4.075 to 3.582). This is because the TFDS fallback logic (ds.shard) performs a "logical shard" that effectively reconstructs the un-shuffled sequential order of the original dataset across the global batch. By physically re-sharding into dedicated files, each host can stream contiguous records into its local shuffle buffer. This ensures that when host data is combined, the global batch contains a truly randomized distribution of data categories, leading to higher-quality gradients and better model performance.

# Tests

Local conversation - [example log](https://paste.googleplex.com/6404668164014080)

Before this change:
* Warning from loading both training and eval dataset, and there is huge performance hit for each training steps
* end-to-end training logs by loading real dataset, and performance is `Tokens/s/device: 80-90` per step [link](https://screenshot.googleplex.com/5AX6KxMnS3RDPBB) 

```
/deps/src/maxtext/input_pipeline/tfds_data_processing.py:64: UserWarning: WARNING: Inefficient dataloading. Your c4/en:3.0.5 contains 128shards, smaller than dataloading_host_count=2048. This is known to lead to inefficient dataloading.

/deps/src/maxtext/input_pipeline/tfds_data_processing.py:64: UserWarning: WARNING: Inefficient dataloading. Your c4/en:3.0.5 contains 64shards, smaller than dataloading_host_count=2048. This is known to lead to inefficient dataloading
```

After this change:
* We reshard the dataset based on host count
* No warning is found in logs: [link](https://cloudlogging.app.goo.gl/T7xyvaKftuYzrFcm6)
* end-to-end training logs by loading real dataset, and performance is `Tokens/s/device: 600-700` per step [link](https://screenshot.googleplex.com/6wCN8pRFdJ39Rov) 

For training loss:
* before the change, initial loss: 5.819
* after the change, initial loss: 5.816
* observe the training efficiency after re-sharding, the llm_loss at step 77 reduce from 4.075 (before resharding) to 3.582 (after resharding). Both train/validation loss are aligned. 
 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tools/data_generation/reshard_tfds.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+     --num_shards 2048 \
+     --split train
+
+For Eval Dataset:
+
+   python3 tools/data_generation/reshard_tfds.py \
+     --src_dir gs://your-bucket/origin_folder \
+     --dst_dir gs://your-bucket/new_folder \
+     --num_shards 2048 \
+     --split validation
```
Comment (aireenmei) [inline tools/data_generation/reshard_tfds.py]: The only difference for training / eval is split name. We can combine the two examples? Also we should num_workers to the example

File: tools/data_generation/reshard_tfds.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    """Helper to lazily initialize a TFRecordWriter for a given target shard."""
+    if shard_idx not in writers:
+      path = get_shard_path(dst_dir, dataset_name, split, shard_idx, total_shards)
+      writers[shard_idx] = tf.io.TFRecordWriter(path)
+      shard_lengths[shard_idx] = 0
+    return writers[shard_idx]
+
+  # Initialize a tf.data.Dataset to read raw bytes from the source TFRecord files.
+  # A large buffer size (32MB) is used to improve I/O throughput, especially on GCS.
+  buffer_size = 32 * 1024 * 1024
```
Comment (aireenmei) [inline tools/data_generation/reshard_tfds.py]: Make this configurable?

File: tools/data_generation/reshard_tfds.py (hunk 1 above)
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: Sounds good, let me combine them as one.

File: tools/data_generation/reshard_tfds.py (hunk 2 above)
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: Sounds good. Let me update.

File: tools/data_generation/reshard_tfds.py (hunk 1 above)
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: Sure, let me add default num_works=16 for efficiency. Testing WIP.

Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: Sanity checked the # of shards in `gs://mlperf-6-submission/tfds-tmp2/c4/en/3.0.5/` with successful local runs: [link](https://paste.googleplex.com/6048441949093888).

Also compared with dataset_info.json, it's aligned with expectation, see [comparison](https://diff.googleplex.com/#key=wjsVZz3l2u5k).

**Review by shuningjin — APPROVED** [review]
Approve to unblock.

File: tools/data_generation/reshard_tfds.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    # Setup multiprocessing pool
+    num_workers = args.num_workers
+
+    # Ensure the target number of shards is divisible by the number of workers
+    # to maintain proper load balancing and deterministic write distributions
+    if args.num_shards % num_workers != 0:
+      for i in range(num_workers, 0, -1):
+        if args.num_shards % i == 0:
+          num_workers = i
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: Let me add a progress bar for this.

File: tools/data_generation/reshard_tfds.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+    with multiprocessing.Pool(processes=num_workers) as pool:
+      results = pool.starmap(reshard_raw_bytes_worker, tasks)
+
+    # Aggregate the results (shard lengths) from all workers
+    all_shard_lengths = {}
+    for r in results:
+      all_shard_lengths.update(r)
+
+    # Verify the total number of examples processed matches the original metadata
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: Updated accordingly.

File: tools/data_generation/reshard_tfds.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py]: After updates, we should see something like this:

```
Worker 0: 37386it [00:08, 5679.60it/s]
Worker 1: 37301it [00:08, 5908.12it/s]
Worker 2: 34547it [00:08, 5358.98it/s]
Worker 3: 37706it [00:08, 5912.57it/s]
Worker 4: 38075it [00:08, 5964.88it/s]
Worker 5: 33327it [00:08, 5321.00it/s]
Worker 6: 35198it [00:08, 5769.77it/s]
Worker 7: 35968it [00:08, 5690.44it/s]
Worker 8: 29798it [00:07, 5954.12it/s]
Worker 9: 35116it [00:08, 5722.14it/s]
Worker 10: 36393it [00:08, 5576.59it/s]
Worker 11: 36385it [00:08, 5655.79it/s]
Worker 12: 36154it [00:08, 6069.33it/s]
Worker 13: 37997it [00:08, 5796.30it/s]
Worker 14: 37509it [00:08, 5958.19it/s]
Worker 15: 32950it [00:08, 5128.82it/s]
Worker 11: 2310it [00:01, 3870.42it/s]
Worker 3: 34699it [00:07, 5382.69it/s]
```

**Review by aireenmei — APPROVED** [review]

File: tools/data_generation/reshard_tfds.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  if not tf.io.gfile.exists(args.dst_dir):
+    tf.io.gfile.makedirs(args.dst_dir)
+
+  target_splits = [s.strip() for s in args.split.split(",") if s.strip()]
+
+  # Load source metadata once
+  print(f"Loading metadata from {args.src_dir}...")
+  info_path = os.path.join(args.src_dir, "dataset_info.json")
+  if not tf.io.gfile.exists(info_path):
+    raise FileNotFoundError(f"Required metadata file not found: {info_path}")
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tools/data_generation/reshard_tfds.py:167]: Updated accordingly.
